db/snapshot_impl.h

```diff
@@ -21,6 +21,10 @@ class SnapshotList;
 class SnapshotImpl : public Snapshot {
  public:
   SequenceNumber number_;  // const after creation
+  // This is used by WritePrepared transactions to limit the scope of queries to
+  // IsInSnpashot. It indicates the smallest uncommitted data at the time the
+  // snapshot was taken.
+  SequenceNumber smallest_prepare_ = 0;
```

Comment:
I think the approach is OK. Maybe all it `all_committed` or something like that, so that it can make sense even if 2pc is not enabled, and set it to the same as `number_` in that case.

Reply:
smallest_prepare_ cannot be const since it is modified after the snapshot object is created. We did not want to modify the exist GetSnapshot code, so the txn db, calls GetSnapshot and then could potentially instrument it by setting smallest_prepare_.

Reply:
how about at least make `number_` a const to denote it never changes? Previously we always pass `const Snapshot*` around so `number_` it is fine was not a const.

Reply:
There is a lot of code relying on number_ being a non-const. We still return a const Snapshot to the user but internally we relaxed the const modified in GetSnapshotImpl as it is not yet returned to the user and further processing should be allowed.